Näkki improvements

LassiKuisma opened the pull request:
Big rework for näkki:
- näkki states are now nodes
- configuration warnings for states and näkki
- create archetypes of näkki that can be placed in levels

TODOs:
- [ ] remove old näkki and replace it with new näkki (?)
- [x] make näkki trigger a `[Tool]`
- [x] code style:
  - [x] shorthand for new empty array
  - [x] `[MustSetInEditor]` for default & current state
  - [x] make näkki's current state public
  - [x] make exported fields public properties
  - [x] mark "prewired" nodes with `[MustSetInEditor]`
  - [x] naming: `_fooBar` means private, otherwise make public property
  - [x] states: add timer with code instead of in editor
  - [x] mark states with `[GlobalClass]` ?
  - [x] `[ExportGroup("Prewire")]`
  - [x] investigate if `notifyPropertyListChanged: true` actually needed
  - [x] organize functions/properties, most important at the top
  - [x] cleanup `WakeUpNakkiTrigger.cs`

LassiKuisma committed:
WIP: näkki rework

LassiKuisma committed:
näkki states: state names

LassiKuisma committed:
make ai base state abstract

LassiKuisma committed:
näkki: stalk state

LassiKuisma committed:
WIP: näkki attack state

LassiKuisma committed:
näkki attack: custom time for attack

LassiKuisma committed:
näkki underwater state

LassiKuisma committed:
näkki state names

LassiKuisma committed:
näkki: after attack dive for a reduced time

LassiKuisma committed:
näkki attack: water splash

LassiKuisma committed:
näkki states: override states

LassiKuisma committed:
näkki states: default values

LassiKuisma committed:
näkki cooldown for dive state

LassiKuisma committed:
näkki tweaks: chance to move to/emerge at player position

LassiKuisma committed:
randomized times for näkki

LassiKuisma committed:
näkki: reset state on player respawn

LassiKuisma committed:
näkki: custom animation speeds

LassiKuisma committed:
fix: näkki reset state on respawn

LassiKuisma committed:
näkki animation fixes

LassiKuisma committed:
näkki: dive on trigger

LassiKuisma committed:
näkki: emerge from water on trigger

LassiKuisma committed:
fix: math stuff

LassiKuisma committed:
näkki archetypes

LassiKuisma committed:
näkki: dive cooldown

LassiKuisma committed:
näkki stalk state: chance to dive

LassiKuisma committed:
näkki code cleanup: DRY

LassiKuisma committed:
refact: DRY when näkki picks new state

LassiKuisma committed:
refact: move näkki variables

LassiKuisma committed:
update näkki's dive state

start detecting player when emerging, and allow movement while emerging

LassiKuisma committed:
fix: näkki archetypes stalk state

LassiKuisma committed:
näkki: player namespace

LassiKuisma committed:
näkki: [Tool] for ai states

LassiKuisma committed:
näkki: [Tool]

LassiKuisma committed:
näkki trigger: [Tool]

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+[Tool]
+public partial class NakkiV2 : Path2D {
+	public override string[] _GetConfigurationWarnings() {
+		var warnings = base._GetConfigurationWarnings() ?? Array.Empty<string>();
```
I think shorthand collection initialization is available due to bump to .NET8
```suggestion
		var warnings = base._GetConfigurationWarnings() ?? [];
```

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+
+		if (_defaultState == null) {
+			warnings = warnings.Append("Default state is not set").ToArray();
+		}
```
We'd get rid of manually implementing this check by adding a `[MustSetInEditor]` to `_defaultState`. Sadly, this doesn't work out of the box (yet).

If you prefer having `_defaultState` as a private nullable field instead of changing it to a public non-nullable property backed by a nullable field, this should be doable with:
1. making the `BindingFlags` changes you talked about earlier (to get access to `private` fields/properties)
2. duplicating the property warning generation code in `CheckCommonConfigurationWarnings()` to process results from `GetFields()` in addition to `GetProperties()`

It might be neat to have a non-nullable `DefaultState` available, though, so I wouldn't mind having it as a public exported `[MustSetInEditor]` property, either.

My agenda is to minimize custom property configuration warning code as much as possible. Either way works, your call.

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+			_currentState = value;
+			CurrentState?.EnterState(this);
+		}
+	}
```
Can't find any best-practices to cite, so this is likely a me-problem, but I find private properties a bit icky. The property is already _practically_ public due to having the publicly available `SwitchToState`, which just calls the setter.

I think it would be cleaner to make `CurrentState` public and removing `SwitchToState` in favour of just using the property/setter directly. Alternatively, the property could be removed and the setter implementation moved to `SwitchToState`. However, regariding the `CurrentState` could-have-not-null-type comment above, I'd prefer exposing the property as public.

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+	[Export] private NakkiAiState? _defaultState;
+
+	private NakkiAiState? _currentState;
+	private NakkiAiState? CurrentState {
```
This is a hacky/odd convention, but a part of the reason why `[MustSetInEditor]` and related utilities exist is the fact that there are a lot of fields/properties which are _effectively non-nullable after `_Ready` has been successfully executed._ Therefore, I prefer to have such "effectively non-nullable" fields/properties with non-nullable type. Yes, this is us blatantly lying to the compiler and making silly errors possible in anything that can be accessed before `_Ready()` executes, with the benefit of reducing a lot of the null-checking mental overhead.

`[MustSetInEditor]` et.al. utilities exist to validate and generate configuration warnings for any properties that should be assigned in-editor, which does not apply here. However, this is a bit similar case in that `_currentState` is always non-nullable after `_Ready()` has exited. Correct me if I read the code wrong, or made some silly assumptions, but assuming `_defaultState` is also set-up properly (remembering that `_defaultState` is effectively `[MustSetInEditor]`), the `_currentState` can never be null after being reset to the default state (which happens in the `_Ready()`).

Therefore, applying my non-nullable-after-ready convention here, `CurrentState` could be marked non-null. E.g. you could drop the `?` from the property type, and the property accessors' read/write ops could be implemented with `GetNotNullExportPropertyWithNullableBackingField` and `SetExportProperty`.

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+	private RayCast2D? _lineOfSight;
+	public float _detectionLevel = 0.0f;
+	[Export] private float _detectionGain = 100.0f;
+	[Export] private float _detectionDecay = 60.0f;
```
Why are these private fields? Could these be updated to public properties? As I said earlier when we discussed the `BindingFlags` thing, I see any `[Export]` as inherently `public`. The only exception I can think of is something that breaks if updated at runtime (so you'd like that to be configurable in-editor but not at runtime.

Looking at code, modifying these from code at runtime is (while a bit silly) completely safe, so I'd just change these to public props.
```suggestion
	[Export]
	private float DetectionGain { get; set; } = 100.0f;
	[Export]
	public float DetectionDecay { get; set; } = 60.0f;
```

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+
+	private Player? _player;
+	private RayCast2D? _lineOfSight;
+	public float _detectionLevel = 0.0f;
```
If this is intentionally public, this should be 
```suggestion
	public float DetectionLevel { get; set; } = 0.0f;
```
If not intentional, mark as private.

Underscore prefix marks a private member field.

ritinae replied:
There are a few other `public X _hasPrivatePrefix;` fields, so same thing for those, too.

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+			return;
+		}
+
+		_nakkiEntity = GetNode<PathFollow2D>("NäkkiEntity");
```
I prefer having any such "prewired" node dependencies as public exported `[MustSetInEditor]` properties. While this results in _a lot_ of boilerplate for complex classes like this, we gain the benefit of configuration warnings in case something is missing or node paths have changed AND just renaming/moving nodes usually don't require code changes.

I assume lot of this init is probably copy-paste from code that predates `[MustSetInEditor]`, but that's one cleanup we could apply here.

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+			_playerIsDead = false;
+			_fakePlayer.Visible = false;
+			_hand.Visible = false;
+			_targetProgress = null;
```
Could these be in `ResetStateToDefault` as I'd assume such reset method to actually reset the state to a consistent default state?

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+
+	public float PathLength() {
+		return Curve.GetBakedLength();
+	}
```
```suggestion
	public float PathLength => Curve.GetBakedLength();
```

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,291 @@
+	}
+
+	public PathFollow2D? _nakkiEntity;
+	private float? _targetProgress;
```
Not a huge fan of nullable primitives, but I think the nullability is a lot easier to get rid of if/when we have a Behaviour Tree implementation in place, so I think its fine for now.

ritinae commented on mobs/nakki_v2/ai_states/NakkiEmergeOnTriggerState.cs:
```diff
@@ -0,0 +1,93 @@
+			return;
+		}
+
+		_emergeOnTimeout = GetNode<Timer>("Timer");
```
Would it make sense creating this child timer with `new Timer()` and `AddChild` instead of `GetNode`?

The timer `WaitTime` is managed here internally and it can't really be looping, so there isn't much that can/should be configured in the editor for the `Timer` node.

And if we didn't have the requirement to pre-wire these child timer nodes in the editor, there really isn't much reason to have e.g. the `nakki_emerge_on_trigger_state.tscn` scenes around, as those scenes would contain only the root-node.

And if we didn't have those state-type scenes around, we could mark the `NakkiSomeState` classes as `[GlobalClass]` so that they are available in editor via the "Add node" dialog. This could make the simpler state nodes a lot more ergonomic to use.

ritinae replied:
It would be really nice to have these timers created on-demand with `GetTree().CreateTimer()`, but those timers don't have a `.Stop()` method, which complicates things.

For cases where the timer is only used to set a boolean, like this state class, it should be fine to just `_timer.TimeLeft = 0f`, though. This makes the timer finish immediately, which sadly sets the boolean, too, but this might not be that bad, mayb. _Maybe_ one-off timers could work. `AddChild` approach might still be cleaner, though :thinking:

ritinae commented on mobs/nakki_v2/ai_states/NakkiEmergeOnTriggerState.cs:
```diff
@@ -0,0 +1,93 @@
+	}
+	private NakkiAiState? _enterStateAfterEmerge;
+
+	[Export] private float _emergeAnimationSpeed = 3.0f;
+	[Export] private float _emergeDelay = 3.0f;
```
I prefer having such stats/configs as public properties.

While encapsulation is nice, I'd prefer open access to configuration by-default. It's a fine line, so I can't really argue better what should be public and what private, but I think have a bit more "configuration"/"stats" variables feel to them, so having them as public might be fine.

As you said, exported private is a bit of an oxymoron, or something 🤷‍♀️

Or yeah, it feels a bit wonky if something somewhere is able to adjust these at runtime, at any moment. On the other hand, there is likely little harm from doing so. Thus, I don't see a reason to prohibit other classes from accessing these. It could be a powerful tool for simple one-off level scripts to be able to adjust stats like these, if needed for some quick hack.

ritinae commented on mobs/nakki_v2/ai_states/NakkiEmergeOnTriggerState.cs:
```diff
@@ -0,0 +1,93 @@
+			.ToArray();
+	}
+
+	[Export]
+	[MustSetInEditor]
```
I usually move the `[MustSetInEditor]` properties as the last exported members of a class, and add a `[ExportGroup("Prewire")]` to the first `[MustSetInEditor]` proprety.

This creates a collapsible "Prewire" group in the inspector, neatly packing all "set-once-and-forget" `MustSetInEditor` properties there. I think the group is additionally collapsed by default, so that really helps keep the inspector clean.

LassiKuisma replied:
I ended up putting exported properties at the top, and clear-current-group (`[ExportGroup("")]`) after them, does this look okay? This way the files are structured like this:

```cs
public override string[] _GetConfigurationWarnings() { ... }

[Export] [ExportGroup("Prewire")] [MustSetInEditor]
public FooBar BarBaz { get => ...; set => ... }

[Export] [MustSetInEditor]
public BooRing FooRing { get => ...; set => ... }


[ExportGroup("")]
[Export] public float FlooBar = 12.34f;

public override void Asd() {
  ...
}
```

To me this seems a lot more clear, you can mentally ignore anything between "get config warnings" and "end of prewire group".

ritinae replied:
TL;DR: This is a really lengthy/wordy way of arguing that I'd prefer if we sticked to the current member ordering where prewire properties are last exported props and there are no magical "clear export group" definitions (the `[ExportGroup("")]`). I somewhat understand the "just ignore the uninteresting parts" argument, but to me placing prewires first feels like placing the most uninteresting part of a class as the first thing that pops up when you glance over the class, which is an antipattern.

I treat the "prewire" properties as "necessary evil" that I hope I can ignore once set up, and never ever think about again. On the other hand, the actual "configuration" properties which I'm allowed/expected to tweak in-editor are the most interesting thing in a class. Therefore, having the "public facing parts", the public exported non-prewire properties as first is the most sensible thing to me. It's easy to glance over the configurable part of the class, by just looking at the topmost exported properties available.

The important thing being: prewire props are uninteresting/unimportant 99% of the time (after the initial setup, that is).

Thus the structure I've been using elsewhere is roughly:
```
class definition {
  [public exported properties]
  [prewire exported properties]
  [private fields]

  [if needed, _GetConfigurationWarnings()]

  [constructor, _Ready(), initializers]

  [public API]
  [other Godot virtual function overrides]
  [private methods]
}
```
I've not been very strict/careful with method ordering for anything below the initializers, but this is roughly the ideal in my headcanon.

In brief: class members are sorted based on how interesting/important they are, in a descending order; the most important thing first. Members are grouped to properties, fields and methods. Properties come before fields, fields come before methods. 

The rough idea is to keep properties/fields at the top, preferring the public/configuration ones to the top. The `_GetConfigurationWarnings()` is grouped with prewire props, as its most closely related to those. Initializer/constructor etc. are related to properly setting up any fields/properties derived from other fields/props, thus they are positioned immediately after property/field definitions.

Then, the next most important thing is the public API of the class; any methods other classes are allowed to access, thus any public methods are best positioned immediately below the initializers. Arguably, Godot virtual overrides are just as important as the public API, so I'm usually unsure which ones to place first. Private/internal APIs/methods are the least interesting to any consumer classes, thus those can be placed last.

> To me this seems a lot more clear, you can mentally ignore anything between "get config warnings" and "end of prewire group".

To me the clarity of the current approach comes from the fact I can easily navigate to either
1. class definition (top of the class), if I'm interested in available configuration properties
2. the `_Ready()`/constructor definition(s), if I'm interested about the public methods available on the class
3. Also, it's easy enough to jump from `[ExportGroup("Prewire")]` to the `_Ready()` method declaration to skip the uninteresting necessary evil in between those.

ritinae commented on mobs/nakki_v2/ai_states/NakkiMovementState.cs:
```diff
@@ -0,0 +1,107 @@
+	[MustSetInEditor]
+	public NakkiStalkState StalkState {
+		get => this.GetNotNullExportPropertyWithNullableBackingField(_stalkState);
+		set => this.SetExportProperty(ref _stalkState, value, notifyPropertyListChanged: true);
```
Is the `notifyPropertyListChanged: true` actually needed for these? I think its only needed for cases where there are some `_ValidateProperty` or `_GetPropertyList` magic going on, or what am I forgetting? :thinking:

ritinae reviewed (commented):
Looking good, some code quality nitpicks and some thoughts :gem:

ritinae commented:
I merged the `cleanup-and-polish` branch as it was effectively a `main` -branch anyway. The main is now somewhat broken state (I bet the game cannot be completed if started from the beginning). This improves over time as the work on the new level transition thingamajig (alongside the new levels themselves) progresses.

This needs to be rebased on top of `main` to resolve the conflicts.

LassiKuisma committed:
add MustSetInEditor for näkki's default state

LassiKuisma committed:
näkki's current state visibility

LassiKuisma committed:
exported fields visibility

LassiKuisma committed:
code style

LassiKuisma committed:
export group for prewired properties

LassiKuisma committed:
use prewired nodes instead of GetNode

LassiKuisma committed:
mark näkki's current state as non-nullable

LassiKuisma committed:
make attack timer prewired + must set

LassiKuisma committed:
create timers with code

ritinae commented:
RE: the todo-item "states: add timer with code instead of in editor"

Due to how Behavior Trees behave, it is preferable if the "action node" / AI state isn't driving using timers, at all. I've yet to crack how to nicely implement support for durations/cooldowns in a clean way to the Behavior Tree system barebones we have in place, but I have a feeling that timers might be a path of pain going forward, unless we can figure out some nice way of supporting those.

The longer version:

The behavior tree does not store the "executing node" between ticks. The current node is determined _each tick_ by _evaluating the whole tree_. That is, we execute nodes one by one, until there are no more nodes to execute. Each node may return `RUNNING`, `SUCCESS` or `FAILURE` depending on if it was able to do anything or not. How the returned result affects the tree evaluation depends on if the node parent is a "sequence" (`BTSequence.cs`) or "selector" (`BTSelector.cs`). If conditions for any nodes executed prior to the node with a timer change, the execution might "unexpectedly" switch to another node, before the timer has a chance to expire. This may or may not be desirable (e.g. for cooldowns this might be completely fine, in some use-cases)

The issue is mostly with anything that uses timers to "do thing X until the timer expires"; Due to how behavior trees work, is is not guaranteed that the node keeps executing until the timer expires. Any node earlier in the tree may grab the execution at any time. Therefore, when the timer expires, the node with the timer is possibly no longer executing, which would break a lot of the current AI states, I'm afraid. Similar issue applies e.g. when waiting for animation to finish before switching states.

All of this might be a non-issue and resolvable with just clever ordering of the nodes. That is, with correctly constructed tree, we can ensure the execution order doesn't contain any such nodes that might grab the execution at unexpected times, ensuring the sequences are fully completed before any other sequences are allowed to execute.

I'm not yet/anymore familiar enough with how Näkki AI works to really go into concrete details how the porting work to behavior trees will look like, so take these as just some random ramblings and/or food for thought. The BT approach is at some parts quite different from how Näkki AI is currently is constructed (the most prevalent difference being that nodes cannot control _when_ the jump to next node happens).

LassiKuisma committed:
[GlobalClass] for näkki states

LassiKuisma committed:
mark ai state base class as [Tool]

Mark the abstract base class with [Tool] so näkki entity can show a warning if its
default state is not set in editor. Default state is marked with [MustSetInEditor],
and for this to work the property must be a [Tool].

LassiKuisma commented:
Reply to @Kailari 's previous comment as GitHub doesn't allow replying directly to it
> RE: the todo-item "states: add timer with code instead of in editor"
> ...
> I'm not yet/anymore familiar enough with how Näkki AI works
> _(rest of message clipped to save space)_

This is how most of the ai states work:
1. Näkki switches to new state and calls its `EnterState()`
2. Upon entering a state: set `doneWithThisState = false;` and start a timer that will set `doneWithThisState = true;`
3. In state's `AiUpdate()` we check if `doneWithThisState == true` and then make näkki to switch to new state

We're not directly switching to a new state with timers, always through a boolean + `AiUpdate()` so that shouldn't be a problem. However, the way I implemented the new timers _(I think I wrote/pushed the code after your comment)_ might create an issue with BT. Since I create timers with `GetTree().CreateTimer(time)` instead of creating a timer node, something like this could happen if we switch state unexpectedly:
1. Näkki enters AbcState
2. AbcState sets `done = false;`, starts a 60 second timer _(timer A)_ that sets `done = true;`
3. For some reason we switch to DefState for some time (less than 60s)
4. We switch back to AbcState, `done = false;` and start new 6s timer _(timer B)_
    - Note: Timer A is still running
6. Timer A timeouts now, setting `done = true;`
7. We exit AbcState earlier than expected

That can be fixed by adding timer nodes in `_Ready()` instead of `GetTree().CreateTimer()`, and when switching to a state we simply restart the timer

LassiKuisma commented on mobs/nakki_v2/ai_states/NakkiAiState.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace Jakojaannos.WisperingWoods;
+
+[Tool]
+public abstract partial class NakkiAiState : Node {
```
@Kailari Any thoughts on `[Tool]` + `abstract class` combo, how bad does this smell? If you try to add `NakkiAiState` as a script to a node you get an error, but Godot doesn't crash or anything:
"Cannot set object script. Script 'res://mobs/nakki_v2/ai_states/NakkiAiState.cs' should not be abstract."

`[Tool]` is needed for näkki's property `[MustSetInEditor] NakkiAiState DefaultState` to show warning if not set

Abstract is nice to get abstract functions.

Can't make it interface as we have `Array<NakkiAiState> PickOneOfThese` which requires the type parameter to be Node.


If this smells too much, we could maybe de-abstract it and add configuration warning, something like
`if you're using NakkiAiState and not a subclass, return warning "Please inherit this class instead of using it directly"`

ritinae replied:
I think the error is just a bit misleading for this case. I think the issue is the other way around; the issue is that abstract classes should never be assigned as object scripts, not that the class should not be abstract. The error message is true in the sense that IF we actually wanted to use the abstract class as a script assigned to a node, _then_ it should not be abstract.

So, I think this is more of a technical limitation of the script file picker than a code smell.

LassiKuisma committed:
näkki trigger code cleanup

LassiKuisma committed:
näkki exported properties: remove 'notifyPropertyListChanged: true'

LassiKuisma committed:
näkki: reorganize code

ritinae commented on Util/RandomNumberGeneratorExtension.cs:
```diff
@@ -54,4 +54,9 @@ public static T PickRandomUnchecked<T>(this RandomNumberGenerator rng, IEnumerab
+	public static float RandomWithVariation(this RandomNumberGenerator rng, float baseChance, float variationPercent) {
+		var v = rng.RandfRange(-variationPercent, variationPercent);
+		return baseChance * (1.0f + v);
+	}
```
I find the naming here is very misleading. The name suggests that this is a method for obtaining a randomized number with some random variation, from which I expected something like `randf(baseChance * randf(-variation, variation))`, not `value * (1 + randf(-variation, variation))`.

As the function seems to not be about obtaining a randomized number with a variation/bias applied, but rather applying a randomized variation to a value, I think e.g. `ApplyRandomVariation` could be more descriptive.

One source of confusion is at least that in the current formula, the `baseChance` is just a base value that gets varied and (when contained within the scope of the function) has nothing to do with chances. I'd suggest renaming it to e.g. just "value".

```suggestion
	public static float ApplyRandomVariation(this RandomNumberGenerator rng, float value, float variationPercent) {
		var variation = rng.RandfRange(-variationPercent, variationPercent);
		return value * (1.0f + variation);
	}
```

ritinae commented on mobs/nakki_v2/NakkiV2.cs:
```diff
@@ -0,0 +1,345 @@
+			value?.EnterState(this);
+		}
+	}
+	private NakkiAiState? _currentState;
```
I strongly dislike the fact the backing field has to be declared nullable, but I guess it cannot be hacked around by just lying to the type system with `private NakkiAiState _currentState = DefaultState;` (are we even allowed to invoke property getters in initializers?) :thinking:

That would then probably still require an odd looking null-coalescing `?.` on non-nullable field when calling `ExitState` (in case `DefaultState` has not been set), or guarding the invocation by null-checking on `_defaultState`, but this probably wouldn't be any cleaner/better than the current implementation.

Cases like this has me questioning the reasoning I had behind creating the `MustSetInEditor` in the first place, although I still think requiring null-checks on stuff that is non-nullable 99% of the time is arguably worse than having odd edge cases like this during initialization.

ritinae commented on mobs/nakki_v2/ai_states/NakkiIdleState.cs:
```diff
@@ -0,0 +1,81 @@
+	private bool _isDoneIdling = false;
+	private RandomNumberGenerator _rng = new();
+
+	public override string[] _GetConfigurationWarnings() {
+		var warnings = base._GetConfigurationWarnings() ?? [];
+
+		if (PickOneOfTheseStatesWhenDoneIdling.Count == 0) {
+			warnings = warnings.Append("Add one or more states to 'pick one of these when done'-list").ToArray();
+		}
+
+		return warnings
+			.Union(this.CheckCommonConfigurationWarnings())
+			.ToArray();
+	}
```
With the recent changes to `CheckCommonConfigurationWarnings` and all the bumps to .NET versions etc, this could now be written as:
```suggestion
	public override string[] _GetConfigurationWarnings() {
		var warnings = PickOneOfTheseStatesWhenDoneIdling.Count == 0
			? ["Add one or more states to 'pick one of these when done'-list"]
			: Arrays.Empty<string>();
		}

		return [
			.. this.CheckCommonConfigurationWarnings(base._GetConfigurationWarnings()),
			.. warnings,
		];
	}
```

ritinae reviewed (approved):
A few more nitpicks etc. but no use in polishing things too much, so I'm happy to merge this as-is if you wish. We can always polish more later when things solidify a bit.

Also sorry how the reviews are so focused on the code style issues. I just want to establish common standards early on, so that we can keep things tidier going forward.

Implementation-wise things are shaping up great, this feels like a good step forwards. There's some heavy reworks still ahead, if we wish to retrofit this to work with behavior trees, I'm afraid. Especially looking at how the state transitions work (`Enter/ExitState` and relying on timers/animations), I'm a bit worried this will surface some flaws with the behavior tree implementation. On the other hand, this offers more use-cases to be explored when extending that functionality, so it should be a good thing in the long run.

LassiKuisma committed:
rename function
